cpp/ql/src/semmle/code/cpp/models/implementations/StdContainer.qll

```diff
@@ -88,6 +88,32 @@ class StdSequenceContainerFrontBack extends TaintFunction {
   }
 }
 
+/**
+ * The standard container function `assign`.
+ */
+class StdSequenceContainerAssign extends TaintFunction {
+  StdSequenceContainerAssign() {
+    this.hasQualifiedName("std", ["vector", "deque", "list", "forward_list"], "assign")
+  }
+
+  /**
+   * Gets the index of a parameter to this function that is a reference to the
+   * value type of the container.
+   */
+  int getAValueTypeParameterIndex() {
+    getParameter(result).getUnspecifiedType() = getDeclaringType().getTemplateArgument(0) // i.e. the `T` of this `std::vector<T>`
+    or
+    getParameter(result).getUnspecifiedType().(ReferenceType).getBaseType() =
+      getDeclaringType().getTemplateArgument(0)
```

Comment:
If we have
```cpp
typedef int myInt;
std::vector<myInt> v;
```

don't we then have an `assign` overload where the parameter 1 `getUnspecifiedType` is `int &` while template argument 0 is `myInt`? Then the types won't match.

Reply:
Good point, I will add a test and fix for this.

Reply:
Do other models have the same problem?

Reply:
I've added a test but it work already without a fix.  I think what's going on is that templates are already built on the underlying type, not the typedef (presumably to avoid unnecessary duplication?).  I'm happy to add `getUnderlyingType()` onto `getTemplateArgument(0)` if you're still concerned about this though.

Reply:
You're right, it's fine as it is. Templates have to do this because there's only one copy of the template for both `vector<int>` and `vector<myInt>`.

Reply:
How about `vector<const int*>`, then? if that's broken, then the fix is to replace `getUnspecifiedType` with `getUnderlyingType`, I think.

Reply:
Happy to add more tests, even if what I have is correct it's clearly not an area I thought too deeply about...

Reply:
I added `const int *` and it didn't work, the only fix that seems to work for both `float` and `const int *` is `getUnspecifiedType()` on both sides on the equation (which I've done).

Reply:
Some of the lines look a bit long now. Does the format check still pass?

Reply:
Ah, I knew I had forgotten something ... fixed.